Row-wise and col-wise vignettes (#4902)
tidyverse/dplyr: vignettes/rowwise.Rmd

gf <- df %>% group_by(g)
rf <- df %>% rowwise(g)
```

If we compute some properties of `y`, you'll notice the results look different:

```{r}
gf %>% mutate(type = typeof(y), length = length(y))
rf %>% mutate(type = typeof(y), length = length(y))
```

Basically what's happening is that when slicing up the columns to pass to the `mutate()` expressions the grouped mutate uses `[` and the rowwise mutate used `[[`. The following code gives a flavour of the different a for loop:

```{r}
# grouped
out1 <- integer(2)
for (i in 1:2) {
  out1[[i]] <- length(df$y[i])
}
out1

# rowwise
out2 <- integer(2)
for (i in 1:2) {
  out2[[i]] <- length(df$y[[i]])
}
out2
```

Note that this magic only applies when you're referring to existing columns, not when you're creating new rows. This is potentially confusing, but we're fairly confident it's the least worst solution.

```{r, error = TRUE}
gf %>% mutate(y2 = y)
rf %>% mutate(y2 = y)
rf %>% mutate(y2 = list(y))
```

### Modelling

`rowwise()` data frames allow you to express solve a variety of modelling problems in what I think is a particularly elegant way. We'll start by using 

```{r}
by_cyl <- mtcars %>% nest_by(cyl)
by_cyl
```

> This is a little different to the usual `group_by()` output: we have visibly changed the structure of the data. Now we three rows (one for each group), and we have a list-col, `data`, that stores the data for that group. Also note that the output is `rowwise()`; this is important because it's going to make working with that list of data frames much easier.

Reviewer: ```suggestion
This is a little different to the usual `group_by()` output: we have visibly changed the structure of the data. Now we have three rows (one for each group), and we have a list-col, `data`, that stores the data for that group. Also note that the output is `rowwise()`; this is important because it's going to make working with that list of data frames much easier.
```

Once we have one data frame per row, it's straightforward to make one model per row:

```{r}
mods <- by_cyl %>% mutate(mod = list(lm(mpg ~ wt, data = data)))
mods
```

And supplement that with one set of predictions per row:

```{r}
mods <- mods %>% mutate(pred = list(predict(mod, data)))
mods
```

You could then summarise the model in a variety of ways:

```{r}
mods %>% summarise(mse = sqrt(mean((pred - data$mpg) ^ 2)))
mods %>% summarise(rsq = summary(mod)$r.squared)
mods %>% summarise(broom::glance(mod))
```

Or easily access the parameters of each model:

```{r}
mods %>% summarise(broom::tidy(mod))
```

## Previously

We've made a number of previous attempts at solving this problem, starting with `do()`. `do()` has been questioning for a long time, because it felt very different to the other dplyr verbs. It had two main modes of operation:

* Without argument names: you could call functions that input and output 
  data frames using `.` to refer to the "current" group. This has been

>   superseeded by the combination of `across()` plus the more permission 

Reviewer: ```suggestion
  superseded by the combination of `across()` plus the more permissive 
```

  `summarise()` which can now create multiple columns and multiple rows.

* With arguments: it worked like `mutate()` but automatically wrapped every
  element in a list. I now believe that is too magical, so can be replaced
  by `mutate()`. (Which now gives a error message with hint to use `list()` 
  in case you forget).

> `rowwise()` was also questioning for quite some time, partly because I didn't appreciate how many people needed the native ability to compute summaries across multiple variables for each row. As an alternative we recommended performing rowwise operations with the purrr `map()` functions. However, this was challenging because you needed to pick a map function based on the number of arguments that were varying and the type of result, which required quite some knowledge of purrr functions. 

Reviewer: ```suggestion
`rowwise()` was also questioning for quite some time, partly because I didn't appreciate how many people needed the native ability to compute summaries across multiple variables for each row. As an alternative, we recommended performing rowwise operations with the purrr `map()` functions. However, this was challenging because you needed to pick a map function based on the number of arguments that were varying and the type of result, which required quite some knowledge of purrr functions. 
```
